dagql/objects.go

```diff
@@ -0,0 +1,857 @@
+	"context"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/dagger/dagger/dagql/idproto"
+	"github.com/iancoleman/strcase"
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+// Class is a class of Object types.
+//
+// The class is defined by a set of fields, which are installed into the class
+// dynamically at runtime.
+type Class[T Typed] struct {
+	inner   T
+	idable  bool
+	fields  map[string]*Field[T]
+	fieldsL *sync.Mutex
+}
+
+type ClassOpts[T Typed] struct {
+	// NoIDs disables the default "id" field and disables the IDType method.
+	NoIDs bool
+
+	// Typed contains the Typed value whose Type() determines the class's type.
+	//
+	// In the simple case, we can just use a zero-value, but it is also allowed
+	// to use a dynamic Typed value.
+	Typed T
+}
+
+// NewClass returns a new empty class for a given type.
+func NewClass[T Typed](opts_ ...ClassOpts[T]) Class[T] {
+	var opts ClassOpts[T]
+	if len(opts_) > 0 {
+		opts = opts_[0]
+	}
+	class := Class[T]{
+		inner:   opts.Typed,
+		fields:  map[string]*Field[T]{},
+		fieldsL: new(sync.Mutex),
+	}
+	if !opts.NoIDs {
+		class.Install(
+			Field[T]{
+				Spec: FieldSpec{
+					Name:        "id",
+					Description: fmt.Sprintf("A unique identifier for this %s.", class.TypeName()),
+					Type:        ID[T]{inner: opts.Typed},
+					Pure:        true,
+				},
+				Func: func(ctx context.Context, self Instance[T], args map[string]Input) (Typed, error) {
+					return NewDynamicID[T](self.ID(), opts.Typed), nil
+				},
+			},
+		)
+		class.idable = true
+	}
+	return class
+}
+
+func (class Class[T]) Typed() Typed {
+	return class.inner
+}
+
+func (class Class[T]) IDType() (IDType, bool) {
+	if class.idable {
+		return ID[T]{inner: class.inner}, true
+	} else {
+		return nil, false
+	}
+}
+
+func (class Class[T]) Field(name string) (Field[T], bool) {
+	class.fieldsL.Lock()
+	defer class.fieldsL.Unlock()
+	field, ok := class.fields[name]
+	if !ok {
+		return Field[T]{}, false
+	}
+	return *field, ok
+}
+
+func (class Class[T]) Install(fields ...Field[T]) {
+	class.fieldsL.Lock()
+	defer class.fieldsL.Unlock()
+	for _, field := range fields {
+		field := field
+		class.fields[field.Spec.Name] = &field
+	}
+}
+
+var _ ObjectType = Class[Typed]{}
+
+func (cls Class[T]) TypeName() string {
+	return cls.inner.Type().Name()
+}
+
+func (cls Class[T]) Extend(spec FieldSpec, fun FieldFunc) {
+	cls.fieldsL.Lock()
+	defer cls.fieldsL.Unlock()
+	cls.fields[spec.Name] = &Field[T]{
+		Spec: spec,
+		Func: func(ctx context.Context, self Instance[T], args map[string]Input) (Typed, error) {
+			return fun(ctx, self, args)
+		},
+	}
+}
+
+// Definition returns the schema definition of the class.
+//
+// The definition is derived from the type name, description, and fields. The
+// type may implement Definitive or Descriptive to provide more information.
+//
+// Each currently defined field is installed on the returned definition.
+func (cls Class[T]) TypeDefinition() *ast.Definition {
```

Comment:
nit: here, and quite a few other places in this file the doc-string refers to the wrong method name. `Definition` => `TypeDefinition`.

Reply:
woops, will patch that up - thanks!